File: GetLocation/GetLocation/ViewController.m
//
//  ViewController.m
//  GetLocationTest
//
//  Created by Oliver Drobnik on 15.05.14.
//  Copyright (c) 2014 Cocoanetics. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <CLLocationManagerDelegate>

@end

@implementation ViewController
{
   CLLocationManager *_locationMgr;
   CLLocation *_mostRecentLoc;
   CLGeocoder *_geoCoder;
   
   NSDictionary *_addressDictionary;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
   
   // initialize geo coder
   _geoCoder = [[CLGeocoder alloc] init];
   
   [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didBecomeActive:) name:UIApplicationDidBecomeActiveNotification object:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
   [super viewWillAppear:animated];
   
   [self _enableLocationUpdatesIfAuthorized];
}

- (void)dealloc
{
   [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)_enableLocationUpdatesIfAuthorized
{
   CLAuthorizationStatus authStatus = [CLLocationManager authorizationStatus];
   
   switch (authStatus)
   {
      default: // also handles iOS 8 status codes
      case kCLAuthorizationStatusNotDetermined:
      case kCLAuthorizationStatusAuthorized:
      {
         NSLog(@"authorized or not determined");
         
         // initialize location manager
         if (!_locationMgr)
         {
            _locationMgr = [[CLLocationManager alloc] init];
            _locationMgr.delegate = self;

#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_7_1
            // on iOS 8 you need to explicitly request authorization
            if ([_locationMgr respondsToSelector:@selector(requestWhenInUseAuthorization)])
            {
               [_locationMgr requestWhenInUseAuthorization];
            }
#endif
            
            [_locationMgr startUpdatingLocation];
         }
         break;
      }
         
      case kCLAuthorizationStatusDenied:
      case kCLAuthorizationStatusRestricted:
      {
         NSLog(@"policy has restricted location updates or user denied it");
         
         _locationMgr = nil;
         break;
      }
   }
}

#pragma mark - Helpers

- (void)_updateLatLongLabelsWithLocation:(CLLocation *)location
{
   dispatch_async(dispatch_get_main_queue(), ^{
      self.latitudeLabel.text = [NSString stringWithFormat:@"%f", location.coordinate.latitude];
      self.longitudeLabel.text = [NSString stringWithFormat:@"%f", location.coordinate.longitude];
   });
}

- (void)_updateAddressLabelWithPlacemark:(CLPlacemark *)placemark
{
   dispatch_async(dispatch_get_main_queue(), ^{
      _addressDictionary = placemark.addressDictionary;
      NSString *addressStr = ABCreateStringWithAddressDictionary(_addressDictionary, YES);
      self.addressTextView.text = addressStr;
   });
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations
{
   CLLocation *location = [locations lastObject];
   
   if (location.coordinate.longitude != _mostRecentLoc.coordinate.longitude ||
       location.coordinate.latitude != _mostRecentLoc.coordinate.latitude)
   {
      [self _updateLatLongLabelsWithLocation:location];
      
      [_geoCoder reverseGeocodeLocation:location completionHandler:^(NSArray *placemarks, NSError *error) {
         if (placemarks)
         {
            CLPlacemark *placemark = [placemarks firstObject];
            [self _updateAddressLabelWithPlacemark:placemark];
         }
         else
         {
            NSLog(@"Error from geocoder: %@", [error localizedDescription]);
         }
      }];
      
      _mostRecentLoc = [locations lastObject];
   }
}

#pragma mark - Notifications

// when app comes into foreground and also following the initial authorization dialog
- (void)didBecomeActive:(NSNotification *)notification
{
   [self _enableLocationUpdatesIfAuthorized];
}

#pragma mark - Actions

- (IBAction)copyAddress:(id)sender
{
   ABRecordRef record = ABPersonCreate();
   
   ABRecordSetValue(record, kABPersonFirstNameProperty, (__bridge CFTypeRef)(@"My Location"), nil);
   
   ABMutableMultiValueRef multiHome = ABMultiValueCreateMutable(kABMultiDictionaryPropertyType);
   
   bool didAddHome = ABMultiValueAddValueAndLabel(multiHome, (__bridge CFTypeRef)(_addressDictionary), kABHomeLabel, NULL);
   
   if(didAddHome)
   {
      ABRecordSetValue(record, kABPersonAddressProperty, multiHome, NULL);
      
      NSLog(@"Address saved.");
   }
   
   NSString *mapString = [NSString stringWithFormat:@"http://maps.apple.com/?q=%f,%f&sspn=0.002828,0.006132&sll=%f,%f", _mostRecentLoc.coordinate.latitude, _mostRecentLoc.coordinate.longitude, _mostRecentLoc.coordinate.latitude, _mostRecentLoc.coordinate.longitude];
   
   // Adding url
   ABMutableMultiValueRef urlMultiValue = ABMultiValueCreateMutable(kABMultiStringPropertyType);
   ABMultiValueAddValueAndLabel(urlMultiValue, (__bridge CFTypeRef)(mapString), (__bridge CFTypeRef)@"Map URL", NULL);
   ABRecordSetValue(record, kABPersonURLProperty, urlMultiValue, nil);
   CFRelease(urlMultiValue);
   
   CFArrayRef people = CFArrayCreate(NULL, &record, 1, NULL);
   
   NSData *data = CFBridgingRelease(ABPersonCreateVCardRepresentationWithPeople(people));
   CFRelease(people);
   CFRelease(record);
   
   NSString* vcardString;
   vcardString = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
   NSLog(@"%@",vcardString);
   
   
   NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
   NSString *documentsDirectory = [paths objectAtIndex:0]; // Get documents directory
   
   NSString *filePath = [documentsDirectory stringByAppendingPathComponent:@"pin.loc.vcf"];
   [vcardString writeToFile:filePath
                 atomically:YES encoding:NSUTF8StringEncoding error:NULL];
   
   NSURL *url =  [NSURL fileURLWithPath:filePath];
   NSLog(@"url> %@ ", [url absoluteString]);
   
   
   // Share Code //
   NSArray *itemsToShare = [[NSArray alloc] initWithObjects: url, nil] ;
   UIActivityViewController *activityVC = [[UIActivityViewController alloc] initWithActivityItems:itemsToShare applicationActivities:nil];
   activityVC.excludedActivityTypes = @[UIActivityTypePrint,
                                        UIActivityTypeCopyToPasteboard,
                                        UIActivityTypeAssignToContact,
                                        UIActivityTypeSaveToCameraRoll,
                                        UIActivityTypePostToWeibo];
   
   if (UI_USER_INTERFACE_IDIOM()==UIUserInterfaceIdiomPhone)
   {
      [self presentViewController:activityVC animated:YES completion:nil];
   }
}


@end

File: GetLocation/GetLocationSwift/ViewController.swift
//
//  ViewController.swift
//  GetLocationSwift
//
//  Created by Geoff Breemer on 04/10/14.
//  Copyright (c) 2014 Cocoanetics. All rights reserved.
//

import UIKit
import CoreLocation
import Dispatch
import AddressBookUI
import Foundation

class ViewController: UIViewController, CLLocationManagerDelegate
{
    @IBOutlet var latitudeLabel: UILabel!
    @IBOutlet var longitudeLabel: UILabel!
    @IBOutlet var addressTextView:  UITextView!
    
    private var _locationMgr: CLLocationManager!
    private var _mostRecentLoc: CLLocation?
    private var _geoCoder : CLGeocoder?
    private var _addressDictionary = [NSObject : AnyObject]()
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        // initialize geo coder
        _geoCoder = CLGeocoder()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector:"didBecomeActive:", name:UIApplicationDidBecomeActiveNotification, object:nil)
    }
    
    override func viewWillAppear(animated: Bool)
    {
        super.viewWillAppear(animated)
        
        _enableLocationUpdatesIfAuthorized()
    }
    
    // MARK: Notifications
    
    deinit
    {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // when app comes into foreground and also following the initial authorization dialog
    func didBecomeActive(notification: NSNotification)
    {
        self._enableLocationUpdatesIfAuthorized()
    }
    
    // MARK: CLLocationManagerDelegate
    
    func locationManager(manager: CLLocationManager!, didUpdateLocations locations: [AnyObject]!)
    {
        var location: CLLocation = locations.last as CLLocation!
        
        if (self._mostRecentLoc == nil || location.coordinate.longitude != self._mostRecentLoc!.coordinate.longitude ||
            location.coordinate.latitude != self._mostRecentLoc!.coordinate.latitude)
        {
            _updateLatLongLabelsWithLocation(location)
            
            _geoCoder?.reverseGeocodeLocation(location, completionHandler: {
                [unowned self]
                (placemarks: [AnyObject]!, error: NSError!) -> Void in
                if (placemarks != nil)
                {
                    var placemark: CLPlacemark = placemarks.first as CLPlacemark
                    self._updateAddressLabelWithPlacemark(placemark)
                }
                else
                {
                    println("Error from geocoder: \(error.localizedDescription)")
                }
                return
            })
            
            self._mostRecentLoc = locations.last as? CLLocation
        }
    }
    
    // MARK: Helpers
    
    func _updateLatLongLabelsWithLocation(location: CLLocation)
    {
        dispatch_async(dispatch_get_main_queue(), {
            [unowned self] in
            self.latitudeLabel?.text = "\(location.coordinate.latitude)"
            self.longitudeLabel?.text = "\(location.coordinate.longitude)"
            return
        })
    }
    
    func _updateAddressLabelWithPlacemark(placemark: CLPlacemark)
    {
        dispatch_async(dispatch_get_main_queue(), {
            [unowned self] in
            self._addressDictionary = placemark.addressDictionary
            var addressStr: String = ABCreateStringWithAddressDictionary(self._addressDictionary, true)
            self.addressTextView?.text = addressStr
        })
    }
    
    func _enableLocationUpdatesIfAuthorized()
    {
        let authStatus = CLLocationManager.authorizationStatus()
        
        if authStatus == CLAuthorizationStatus.Restricted || authStatus == CLAuthorizationStatus.Denied
        {
            println("policy has restricted location updates or user denied it")
            _locationMgr = nil
            return
        }
        
        println("authorized or not determined")
        
        // initialize location manager
        if _locationMgr == nil
        {
            _locationMgr = CLLocationManager()
            _locationMgr.delegate = self

            // on iOS 8 you need to explicitly request authorization
            let iosVersion = NSOperatingSystemVersion(majorVersion: 8, minorVersion: 0, patchVersion: 0)
            if NSProcessInfo.processInfo().isOperatingSystemAtLeastVersion(iosVersion)
            {
                _locationMgr?.requestWhenInUseAuthorization()
            }
            
            _locationMgr.startUpdatingLocation()
        }
    }
    
    // MARK: Actions
    
    @IBAction func copyAddress(sender: AnyObject)
    {
        var record: ABRecordRef = ABPersonCreate().takeRetainedValue()
        ABRecordSetValue(record, kABPersonFirstNameProperty, "My Location", nil)
        
        var multiHome: ABMutableMultiValueRef = ABMultiValueCreateMutable(ABPropertyType(kABMultiDictionaryPropertyType)).takeRetainedValue()
        var didAddHome: Bool = ABMultiValueAddValueAndLabel(multiHome, _addressDictionary, kABHomeLabel, nil)
        
        if(didAddHome)
        {
            ABRecordSetValue(record, kABPersonAddressProperty, multiHome, nil)
            println("Address saved.")
        }
        
        var mapString:String = NSString(format: "http://maps.apple.com/?q=%f,%f&sspn=0.002828,0.006132&sll=%f,%f",
            _mostRecentLoc!.coordinate.latitude,
            _mostRecentLoc!.coordinate.longitude,
            _mostRecentLoc!.coordinate.latitude,
            _mostRecentLoc!.coordinate.longitude)
        
        // Adding url
        var urlMultiValue: ABMutableMultiValueRef = ABMultiValueCreateMutable(ABPropertyType(kABMultiStringPropertyType)).takeRetainedValue()
        ABMultiValueAddValueAndLabel(urlMultiValue, mapString, "Map URL", nil)
        ABRecordSetValue(record, kABPersonURLProperty, urlMultiValue, nil)
        
        // Obtain vCard
        var people : NSArray = [record]
        var data: NSData = ABPersonCreateVCardRepresentationWithPeople(people as CFArray).takeRetainedValue()
        var vcardString: String = NSString(data: data, encoding: NSASCIIStringEncoding)
        
        println(vcardString)
        
        // Write vCard
        var paths: [AnyObject] = NSSearchPathForDirectoriesInDomains(.DocumentDirectory, .UserDomainMask, true)
        var documentsDirectory: String = paths.first as String // Get documents directory
        var filePath: String = documentsDirectory.stringByAppendingPathComponent("pin.loc.vcf")
        vcardString.writeToFile(filePath, atomically: true, encoding: NSUTF8StringEncoding, error: nil)
        
        var url: NSURL = NSURL(string: filePath)
        
        println("url> \(url.absoluteString)")
        
        // Share Code
        var itemsToShare: [NSURL] = [url]
        var activityVC: UIActivityViewController = UIActivityViewController(activityItems: itemsToShare, applicationActivities: nil)
        activityVC.excludedActivityTypes = [UIActivityTypePrint,
            UIActivityTypeCopyToPasteboard,
            UIActivityTypeAssignToContact,
            UIActivityTypeSaveToCameraRoll,
            UIActivityTypePostToWeibo]
        
        if (UIDevice.currentDevice().userInterfaceIdiom == .Phone)
        {
            self.presentViewController(activityVC, animated: true, nil)
        }
    }
}